### PR TITLE
Issue/3131 compile reports endpoint

### LINKED ISSUE #3131: Create endpoint to list compile runs
A mix of v1 and v2 endpoints exists for compilation data and compile queues. The goal of this ticket is to create an endpoint that allows you to list the compile queue and the list of compiles.

This part of the data model is partially typed and there is already a `/compiledata` v2 endpoint but it returns only partial data. The typed part constists of data.Compile and data.CompileData. The untyped part is the Reports.

The goal here is to create an endpoint that allows you to list all Compiles in the database. (/v1/compilequeue is similar but only allows to list the compiles that are still queued). The task consists off:

1. List compiles under a new `/v2/compilereport` endpoint. This returns List[data.CompileRun]`
2. Add additional fields to the CompileRun dto:
    - completed
    - success
    - version -> produced by the export of this compile
 3. This list should not return the compile_data field
 4. Allow paging on the `requested` field
 5. Allow sorting on the `requested` field
 
For filtering the following should be supported:

1. Time range filter on requested field
2. not null filter on started field
3. not null filter on completed field
2. boolean filter on success field

### PR DESCRIPTION
**Pull request by andraskvr**
# Description

closes #3131

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

**Comment by andraskvr** [comment]
I've used the name `CompileReport` everywhere, but since there is a different kind of `Report` this might not be ideal. The name `CompileRun` has already been used. I'm open to other naming suggestions

File: src/inmanta/server/validate_filter.py [hunk 1]
```diff
@@ -183,3 +185,64 @@ def process_filters(self, filter: Dict[str, List[str]]) -> Dict[str, QueryFilter
...
+        validated_filter: CompileReportFilterModel = self.validate_filters(filter)
+        query: Dict[str, QueryFilter] = {}
+        if validated_filter.requested:
+            query["requested"] = (QueryType.RANGE, validated_filter.requested)
+        if validated_filter.success is not None:
+            query["success"] = (QueryType.EQUALS, validated_filter.success)
+        if validated_filter.started is not None:
+            query["started"] = (QueryType.IS_NOT_NULL, None) if validated_filter.started else (QueryType.EQUALS, None)
+        if validated_filter.completed is not None:
+            query["completed"] = (QueryType.IS_NOT_NULL, None) if validated_filter.completed else (QueryType.EQUALS, None)
```
Comment (arnaudsjs) [inline src/inmanta/server/validate_filter.py:246]: Why use `IS_NOT_NULL` here? This doesn't match with the API documentation.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/data/__init__.py [hunk 2]
```diff
@@ -2191,6 +2217,102 @@ async def delete_older_than(
...
+        limit: int,
+        environment: uuid.UUID,
+        first_id: Optional[uuid.UUID] = None,
+        last_id: Optional[uuid.UUID] = None,
+        start: Optional[datetime.datetime] = None,
+        end: Optional[datetime.datetime] = None,
+        connection: Optional[asyncpg.connection.Connection] = None,
+        **query: Tuple[QueryType, object],
+    ) -> List[m.CompileReport]:
+        order_by_column = database_order.get_order_by_db_column_name()
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:2258]: Stylistically, I think we should search for a way of improving this filtering/paging framework because

1. all pull requests have a lot of similar code, which suggest duplication
2. this code is quite hard to review, the abstractions of the methods and classes that are used are hard to follow. 

I'll try to give some more detailed comments with specific parts.


I wonder if a query builder class would help, that hides all the little mechanics of counting values and merging queries?

Comment (arnaudsjs) [reply src/inmanta/data/__init__.py:2258]: I agree it's hard to review and improvements are welcome.

File: src/inmanta/server/validate_filter.py (hunk 1 above)
Comment (andraskvr) [reply src/inmanta/server/validate_filter.py:246]: Could you elaborate? 
For the `started` and `completed` values we only need the `IS_NOT_NULL` filter, as the ticket described it, and I don't think the API docs suggest otherwise

Comment (arnaudsjs) [reply src/inmanta/server/validate_filter.py:246]: You are right. You can ignore this comment.

File: src/inmanta/protocol/methods_v2.py [hunk 3]
```diff
@@ -657,3 +657,57 @@ def resource_logs(
...
+    :param limit: Limit the number of instances that are returned
+     :param first_id: The id to use as a continuation token for paging, in combination with the 'start' value,
+            because the order by column might contain non-unique values
+    :param last_id: The id to use as a continuation token for paging, in combination with the 'end' value,
+            because the order by column might contain non-unique values
+    :param start: The lower limit for the order by column (exclusive).
+                Only one of 'start' and 'end' should be specified at the same time.
+    :param end: The upper limit for the order by column (exclusive).
+                Only one of 'start' and 'end' should be specified at the same time.
+    :param filter: Filter the list of returned compile reports.
```
Comment (wouterdb) [inline src/inmanta/protocol/methods_v2.py:685]: I would add the overview of the filtering documentation somewhere centrally. 

It is OK to have it here, but if there is a general approach to filtering, that should be documented somewhere as well.

File: src/inmanta/data/paging.py [hunk 4]
```diff
@@ -128,6 +143,27 @@ async def count_items_for_paging(
         return PagingCounts(total=result[0]["count_total"], before=result[0]["count_before"], after=result[0]["count_after"])
 
 
+class CompileReportPagingCountsProvider(PagingCountsProvider):
+    def __init__(self, data_class: Type[Compile]) -> None:
```
Comment (wouterdb) [inline src/inmanta/data/paging.py:147]: Why is this a parameter at all? afaik there is only one `Type[Compile]`, so it can just as well be hardcoded. 


I think in fact, this could be

```python
class CompileReportPagingCountsProvider(PagingCountsProvider):

    async def count_items_for_paging(
        self,
        environment: uuid.UUID,
        database_order: DatabaseOrder,
        first_id: Optional[Union[uuid.UUID, str]] = None,
        last_id: Optional[Union[uuid.UUID, str]] = None,
        start: Optional[Any] = None,
        end: Optional[Any] = None,
        **query: Any,
      ) -> PagingCounts:
          return await Compile.count_items_for_paging(...)
          
```

Or even have `Compile` extend `PagingCountsProvider` itself

File: src/inmanta/data/__init__.py [hunk 5]
```diff
@@ -2666,7 +2788,9 @@ def to_dto(self) -> m.ResourceAction:
 class ResourceOrder(DatabaseOrder):
     """ Represents the ordering by which resources should be sorted"""
 
-    valid_sort_pattern: Pattern = re.compile("^(resource_type|agent|status|resource_id_value)\\.(asc|desc)$", re.IGNORECASE)
+    valid_sort_pattern: Pattern[str] = re.compile(
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:2791]: I would prefer to simply declare a list of allowed fields, and derive everything from that.

File: src/inmanta/data/__init__.py [hunk 6]
```diff
@@ -2696,7 +2820,7 @@ def _should_be_treated_as_string(self) -> bool:
 class ResourceHistoryOrder(DatabaseOrder):
     """ Represents the ordering by which resource history should be sorted"""
 
-    valid_sort_pattern: Pattern = re.compile("^(date)\\.(asc|desc)$", re.IGNORECASE)
+    valid_sort_pattern: Pattern[str] = re.compile("^(date)\\.(asc|desc)$", re.IGNORECASE)
 
     @classmethod
     def parse_from_string(
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:2826]: Why is this method duplicated?

File: src/inmanta/server/validate_filter.py [hunk 7]
```diff
@@ -183,3 +185,64 @@ def process_filters(self, filter: Dict[str, List[str]]) -> Dict[str, QueryFilter
...
+    def _success_single(cls, v: object) -> object:
+        return parse_single(v, "success")
+
+    @validator("requested", pre=True)
+    @classmethod
+    def parse_requested(cls, v: object) -> List[Tuple[RangeOperator, datetime.datetime]]:
+        return parse_range_operator(v)
+
+
+class CompileReportFilterValidator(FilterValidator[CompileReportFilterModel]):
```
Comment (wouterdb) [inline src/inmanta/server/validate_filter.py:231]: We could make this a lot more declarative by having a dict, that indicates for every field what type of filter it is

```python

class CompileReportFilterValidator(FilterValidator[CompileReportFilterModel]):
     allowed_filter = {
                  "requested": DateRangeFilter,
                  "success": BooleanEqualityFilter,
                  "started": BooleanNoneIsFalse,
                  "completed":BooleanNoneIsFalse,
     }
     
``` 

and validate purely on that

**Review by wouterdb — CHANGES_REQUESTED** [review]

File: src/inmanta/data/__init__.py [hunk 8]
```diff
@@ -2191,6 +2217,102 @@ async def delete_older_than(
...
+        else:
+            db_query += f" ORDER BY {order_by_column} {order}, id {order}"
+        if limit is not None:
+            if limit > DBLIMIT:
+                raise InvalidQueryParameter(f"Limit cannot be bigger than {DBLIMIT}, got {limit}")
+            elif limit > 0:
+                db_query += " LIMIT " + str(limit)
+
+        if backward_paging:
+            db_query = f"""SELECT * FROM ({db_query}) AS matching_records
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/data/__init__.py:2291]: and perhaps even 

```
page_count = base_query.page_count(full_order, [(start,end), (start_id,end_id)])
```

File: src/inmanta/data/__init__.py (hunk 2 above)
Comment (andraskvr) [reply src/inmanta/data/__init__.py:2258]: Yes, I've been thinking about some refactoring for a while, but wasn't sure when to start.
I'll mark this as a draft while I start implementing the suggestions

**Comment by andraskvr** [comment]
@wouterdb  I've made some changes according to your suggestions, but I haven't changed everything yet to use the new classes. Could you take a look?
If the direction is ok, I'll change the rest of the related code (from other queries) to fit it.

File: src/inmanta/data/__init__.py [hunk 9]
```diff
@@ -4031,6 +4093,153 @@ def to_dict(self) -> JsonType:
...
+        self.values = list(values)
+        if initial_filter_statements:
+            self.sql_query += BaseDocument._join_filter_statements(initial_filter_statements)
+        self.filter_started = bool(initial_filter_statements)
+
+    @property
+    def offset(self):
+        return len(self.values) + 1
+
+    def filter(self, filter_statements: List[str], values: List[object]) -> "QueryBuilder":
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:4178]: I would make this class immutable and create a new QueryBuilder on each modification, to make them composable

File: src/inmanta/server/validate_filter.py [hunk 10]
```diff
@@ -34,89 +34,141 @@ def __init__(self, message: str, *args: object) -> None:
...
-        """
-        Processes filters and returns a structured query filter object.
+    def to_query_type(self) -> Optional[Tuple[QueryType, object]]:
+        pass
 
-        :raises InvalidFilter: The supplied filter is invalid.
-        """
-        raise NotImplementedError()
 
+class BooleanEqualityFilter(BaseModel, Filter):
```
Comment (wouterdb) [inline src/inmanta/server/validate_filter.py:95]: Why does this extend BaseModel?

This could use some documenation.

File: src/inmanta/server/validate_filter.py [hunk 11]
```diff
@@ -34,89 +34,141 @@ def __init__(self, message: str, *args: object) -> None:
...
+    field: Optional[bool]
```
Comment (wouterdb) [inline src/inmanta/server/validate_filter.py:96]: I would expect this to be the name of a field, so not a boolean?

File: src/inmanta/data/__init__.py [hunk 12]
```diff
@@ -4031,6 +4093,153 @@ def to_dict(self) -> JsonType:
...
+                self.sql_query += " LIMIT " + str(limit)
+
+        if backward_paging:
+            self.sql_query = f"""SELECT * FROM ({self.sql_query}) AS matching_records
+                                ORDER BY matching_records.{order_by_column} {order},
+                                         matching_records.{db_order.id_column} {order}"""
+
+        return self
+
+    def page_count(
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:4212]: This method here will only produce correct results if the base query has a very particular form (i.e.`SELECT ...` is missing. )

This kind of defeats the purpose of a query builder as an abstraction to handle the complexity of building queries.

I would split the query builder in two types:

- OpaqueQueryBuilder, which works like this one
- A QueryBuilder that is a subclass, that keeps all the parts of the query separated (projection, filter, order, table)

The QueryBuilder can support page_count, the OpaqueQueryBuilder can not.
complex things like order_and_limit always return OpaqueQueryBuilder

I'm going to ask @bartv for a second opinion on this.

File: src/inmanta/data/__init__.py [hunk 13]
```diff
@@ -4031,6 +4093,153 @@ def to_dict(self) -> JsonType:
...
+    @classmethod
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "CompileReportOrder":
+        database_order = super().parse_from_string(sort)
+        return CompileReportOrder(database_order.order_by_column, database_order.order)
+
+
+class QueryBuilder:
```
Comment (bartv) [inline src/inmanta/data/__init__.py:4166]: For me the first issues for reviewing this is the total lack of any comments. This is a helper class so that means that:

- Its purpose should be documented
- Every single method needs documentation on its arguments and return values and preferably a single sentence description of its purpose.

File: src/inmanta/data/__init__.py (hunk 12 above)
Comment (bartv) [reply src/inmanta/data/__init__.py:4212]: Arent we getting to the point that a class like this should symbolicly build the actual query and then a render method that actuall generates the query? I have the feeling a lot of complexity comes from the fact that each method adds to a string which is a irreversible operation.

But not entirely sure, I do not have enough feeling with the code yet.

File: src/inmanta/data/__init__.py (hunk 13 above)
Comment (bartv) [reply src/inmanta/data/__init__.py:4166]: This data module/package is already a big mess I am aware of this. However, when we added helpers this should at least be at the top of the file so that the structure is maintained. Even better would be adding it to a helpers.py

This looks like a small issue, but once a file reaches this size just the size alone makes it complex to review.

File: src/inmanta/data/__init__.py [hunk 14]
```diff
@@ -2191,6 +2256,78 @@ async def delete_older_than(
...
+        environment: uuid.UUID,
+        database_order: DatabaseOrder,
+        first_id: Optional[uuid.UUID] = None,
+        last_id: Optional[uuid.UUID] = None,
+        start: Optional[Any] = None,
+        end: Optional[Any] = None,
+        **query: Any,
+    ) -> PagingCounts:
+        qb = QueryBuilder(
+            f"""FROM {cls.table_name()}
```
Comment (bartv) [inline src/inmanta/data/__init__.py:2271]: Does this even work? Where is the SELECT?

File: src/inmanta/data/__init__.py [hunk 15]
```diff
@@ -2191,6 +2256,78 @@ async def delete_older_than(
...
         await cls._execute_query(query, oldest_retained_date, connection=connection)
 
+    @classmethod
+    async def count_items_for_paging(
+        cls,
+        environment: uuid.UUID,
+        database_order: DatabaseOrder,
+        first_id: Optional[uuid.UUID] = None,
+        last_id: Optional[uuid.UUID] = None,
+        start: Optional[Any] = None,
```
Comment (bartv) [inline src/inmanta/data/__init__.py:2266]: Are you sure this is Any?

File: src/inmanta/data/__init__.py [hunk 16]
```diff
@@ -213,6 +228,48 @@ def ensure_boundary_type(
         else:
             return order_by_column_value
 
+    @property
+    def id_column(self):
+        return "id"
+
+    def as_start_filter(
```
Comment (bartv) [inline src/inmanta/data/__init__.py:235]: This needs a comment and this is way to much copy pasting. I had place it in seperate files and use diff to see the difference except for the slightly different variable names.

**Review by bartv — CHANGES_REQUESTED** [review]
My first high level feedback is that the code structure makes it indeed really difficult to review:

- There is not documentation at all
- The entire querybuilder should not be added to the existing 4000 lines of code
- It looks like a lot of copy paste with very little difference (to the point that I needed to use a differ to actually see the difference)

File: src/inmanta/data/__init__.py (hunk 12 above)
Comment (wouterdb) [reply src/inmanta/data/__init__.py:4212]: I think we are thinking along similar lines

**Comment by arnaudsjs** [comment]
I am going to wait with my review until the big open questions regarding the new query framework have been resolved.

**Comment by bartv** [comment]
> I am going to wait with my review until the big open questions regarding the new query framework have been resolved.

I agree that we should improve how the queries are built, however I think there is a lot more wrong with the data layer that needs to be handled and this would bring us way to far within the scope of a PR. So my suggestions is:

1. Code structure and documentation for the new code has to be improved in this PR
2. Wouter and me will start preparing an epic that handles a number of problems: core vs lsm, orm typing, query building, ...

**Comment by andraskvr** [comment]
I agree that it's getting too big for this one PR, and it's probably best to have a plan on how to do it, so an epic sounds like a good idea.

File: src/inmanta/data/__init__.py [hunk 17]
```diff
@@ -155,19 +155,39 @@ def __init__(self, message: str, *args) -> None:
         self.message = message
 
 
-class DatabaseOrder(metaclass=ABCMeta):
+class DatabaseOrder:
     """Represents an ordering for database queries"""
 
+    valid_sort_columns: Dict[str, Union[Type[datetime.datetime], Type[int], Type[str]]] = {}
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:161]: This field should be documented

File: src/inmanta/data/__init__.py [hunk 18]
```diff
@@ -155,19 +155,39 @@ def __init__(self, message: str, *args) -> None:
...
+    @classmethod
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "DatabaseOrder":
+        valid_sort_pattern: Pattern[str] = re.compile(f"^({'|'.join(cls.valid_sort_columns)})\\.(asc|desc)$", re.IGNORECASE)
+        match = valid_sort_pattern.match(sort)
+        if match and len(match.groups()) == 2:
+            order_by_column = match.groups()[0].lower()
+            validated_order_by_column, validated_order = cls.validator_dataclass()._validate_order_strict(
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:177]: This validation checks the name, but not the type. 

I wonder if it would not be more interesting to make a testcase that goes over every subclass of DatabaseOrder and verifies if `valid_sort_columns` is always valid wrt to the `BaseDocument`

File: src/inmanta/data/__init__.py [hunk 19]
```diff
@@ -213,6 +233,296 @@ def ensure_boundary_type(
...
+    @abstractmethod
+    def build(self) -> Tuple[str, List[object]]:
+        """ Builds up the full query string, and the parametrized value list, ready to be executed """
+        pass
+
+
+class SimpleQueryBuilder(BaseQueryBuilder):
+    """ A query builder suitable for most queries """
+
+    def __init__(
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:410]: All arguments need documentation

File: src/inmanta/data/__init__.py (hunk 9 above)
Comment (wouterdb) [reply src/inmanta/data/__init__.py]: This is still very relevant

File: src/inmanta/data/__init__.py (hunk 18 above)
Comment (andraskvr) [reply src/inmanta/data/__init__.py:180]: In that case we could derive the types from the `__fields` of the specific `BaseDocument` (and `valid_sort_columns` could be only a list of the names), but since we also have cases where we sort on nested columns (for the `ResourceLog`s), that would also require specific tweaking only for that use case.
For the time being I think it's an acceptable compromise that if a developer mixes up the types here, the tests would fail.

In the longer term, for the refactoring epic, I agree that this information should come from one source only.

**Review by wouterdb — CHANGES_REQUESTED** [review]
ok, to separate the long term from the short term:

issues to adress now:

- address the unresolved issues (mostly related to having more comments in the code)

issues to address later
- the overall framework structure

File: src/inmanta/data/__init__.py [hunk 20]
```diff
@@ -213,6 +233,296 @@ def ensure_boundary_type(
...
+
+    @classmethod
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "ResourceOrder":
+        database_order = super().parse_from_string(sort)
+        return ResourceOrder(database_order.order_by_column, database_order.order)
+
+    def get_order_by_db_column_name(self) -> ColumnNameStr:
```
[the remark this answers is not in the document]
Comment (andraskvr) [reply src/inmanta/data/__init__.py]: The columns sometimes have to be cast to a different type (for example casting an `enum` to `varchar` to sort alphabetically instead of the order of the enum), this method is used for this case, while in the API simply the name is used, since the use in the database is an implementation detail from that perspective. 
I renamed these methods to `get_order_by_column_db_name` and `get_order_by_column_api_name` respectively, to make this more explicit.

(This can be e.g. extracted to a separate class later)

File: src/inmanta/server/validate_filter.py (hunk 11 above)
Comment (andraskvr) [reply src/inmanta/server/validate_filter.py:107]: The `Filter` subclasses validate the provided filters with pydantic and describe how to convert them to a `QueryType` which can be used in a database query. (I added this to the docstring)

This means that `field` receives the value to be validated (and the name of the field is set in the `allowed_filters` mapping of the `FilterValidator` classes). This can be simplified when we won't need to support python 3.6 anymore, with generic models

File: src/inmanta/data/__init__.py (hunk 17 above)
Comment (andraskvr) [reply src/inmanta/data/__init__.py:161]: Added it

File: src/inmanta/data/__init__.py (hunk 19 above)
Comment (andraskvr) [reply src/inmanta/data/__init__.py:392]: Added it in the docstring of the `__init__` method

**Review by wouterdb — APPROVED** [review]

File: src/inmanta/protocol/methods_v2.py [hunk 21]
```diff
@@ -657,3 +657,57 @@ def resource_logs(
...
+                Filters should be specified with the syntax `?filter.<filter_key>=value`,
+                for example `?filter.success=True`
+                It's also possible to provide multiple values for the same filter, in this case resources are returned,
+                if they match any of these filter values.
+                For example: `?filter.requested=ge:2021-08-18T09:21:30.568353&filter.requested=lt:2021-08-18T10:21:30.568353`
+                returns compile reports that were requested between the specified dates.
+                Multiple different filters narrow the results however (they are treated as an 'AND' operator).
+                For example `filter.success=True&filter.completed=True` returns compile reports
```
Comment (arnaudsjs) [inline src/inmanta/protocol/methods_v2.py:693]:
```suggestion
                For example `?filter.success=True&filter.completed=True` returns compile reports
```

File: src/inmanta/protocol/methods_v2.py [hunk 22]
```diff
@@ -657,3 +657,57 @@ def resource_logs(
...
+                started: whether the compile has been started or not
+                completed: whether the compile has been completed or not
+                requested: return the logs matching the timestamp constraints. Valid constraints are of the form
+                    "<lt|le|gt|ge>:<x>". The expected format is YYYY-MM-DDTHH:mm:ss.ssssss, so an ISO-8601 datetime string,
+                    in UTC timezone. Specifying microseconds is optional. For example:
+                    `?filter.requested=ge:2021-08-18T09:21:30.568353&filter.requested=lt:2021-08-18T10:21:30`.
+                    Multiple constraints can be specified, in which case only compile reports that match all constraints will be
+                    returned.
+    :param sort: Return the results sorted according to the parameter value.
+                It should follow the pattern `<attribute_to_sort_by>.<order>`, for example `timestamp.desc`
```
Comment (arnaudsjs) [inline src/inmanta/protocol/methods_v2.py:706]:
```suggestion
                It should follow the pattern `?<attribute_to_sort_by>.<order>`, for example `?timestamp.desc`
```

File: src/inmanta/server/services/compilerservice.py [hunk 23]
```diff
@@ -719,3 +723,63 @@ async def get_compile_queue(self, env: data.Environment) -> List[model.CompileRu
...
+        start: Optional[datetime.datetime] = None,
+        end: Optional[datetime.datetime] = None,
+        filter: Optional[Dict[str, List[str]]] = None,
+        sort: str = "requested.desc",
+    ) -> ReturnValue[List[model.CompileReport]]:
+
+        if limit is None:
+            limit = APILIMIT
+        elif limit > APILIMIT:
+            raise BadRequest(f"limit parameter can not exceed {APILIMIT}, got {limit}.")
```
Comment (arnaudsjs) [inline src/inmanta/server/services/compilerservice.py:743]: This code is present in many API endpoints. Maybe we should move this check into `protocol` at some point. @wouterdb That could be part of a refactoring ticket as well.

**Comment by andraskvr** [comment]
@arnaudsjs @bartv could you take a look? 
(Wrt the original ticket, as we discussed to reduce the scope refactoring in this PR and address it later in an epic)

**Comment by arnaudsjs** [comment]
> @arnaudsjs @bartv could you take a look?
> (Wrt the original ticket, as we discussed to reduce the scope refactoring in this PR and address it later in an epic)

Yes. Working on it now.

File: src/inmanta/server/validate_filter.py [hunk 24]
```diff
@@ -131,55 +208,90 @@ def _deleted_single(cls, v: object) -> object:
...
+class FilterValidator(ABC):
+    """
+    This class provides methods to validate and process filters as received via the API.
+    """
+
+    @property
+    @abstractmethod
+    def allowed_filters(self) -> Dict[str, Type[Filter]]:
+        """ A dictionary that determines the mapping between the allowed filters and how they should be parsed and validated"""
+        pass
```
Comment (arnaudsjs) [inline src/inmanta/server/validate_filter.py:229]:
```suggestion
        raise NotImplementedError()
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/data/__init__.py [hunk 25]
```diff
@@ -213,6 +234,326 @@ def ensure_boundary_type(
...
+            self.db_order,
+            self.limit,
+            self.backward_paging,
+        )
+
+    def build(self) -> Tuple[str, List[object]]:
+        if not self.select_clause:
+            raise InvalidQueryParameter("A valid query must have a SELECT clause")
+        full_query = f"""{self.select_clause}
+                         {self._from_clause if self._from_clause else ""}
```
Comment (arnaudsjs) [inline src/inmanta/data/__init__.py:464]: A SQL query without a FROM clause is not going to work, is it? Or am I missing something?

File: src/inmanta/data/__init__.py [hunk 26]
```diff
@@ -213,6 +234,326 @@ def ensure_boundary_type(
...
+    def filter(self, filter_statements: List[str], values: List[object]) -> "PageCountQueryBuilder":
+        return PageCountQueryBuilder(
+            self.select_clause, self._from_clause, self.filter_statements + filter_statements, self.values + values
+        )
+
+    def build(self) -> Tuple[str, List[object]]:
+        if not self.select_clause:
+            raise InvalidQueryParameter("A valid query must have a SELECT clause")
+        full_query = f"""{self.select_clause}
+                         {self._from_clause if self._from_clause else ""}
```
Comment (arnaudsjs) [inline src/inmanta/data/__init__.py:552]: Same here.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/data/__init__.py (hunk 25 above)
Comment (andraskvr) [reply src/inmanta/data/__init__.py:464]: It's technically allowed by PostgreSQL, but you're probably right that it's such a particular use case that in most cases it's better to raise an exception (and in case someone needs a query that allows not setting it, that can be another subclass). I'll make the changes

**Review by arnaudsjs — APPROVED** [review]

**Review by bartv — APPROVED** [review]
